<center>
<h1> TP-Projet d'optimisation numérique </h1>
<h1> Année 2020-2021 - 2e année département Sciences du Numérique </h1>
<h1> Nom: HUC-LHUILLERY  </h1>
<h1> Prénom: Alexia </h1>    
</center>

# Algorithme de Newton
## Implémentation 
 
1. Coder l’algorithme de Newton local tel que décrit dans la section *Algorithme de Newton* (fichier `Algorithme_De_Newton.jl`)

2. Tester l’algorithme sur les fonctions $f_{1}$ , $f_{2}$ avec les points initiaux $x_{011}$ , $x_{012}$ (pour $f_{1}$ ) et $x_{021}$ , $x_{022}$ , $x_{023}$ (pour $f_{2}$ ) donnés en Annexe A.

In [1]:
using Pkg
Pkg.status()
#Pkg.add("DifferentialEquations")

using Markdown
using Test
using LinearAlgebra


      Status `~/.julia/environments/v1.6/Project.toml`
  [7073ff75] IJulia v1.23.2
  [91a5bcdd] Plots v1.23.2
  [37e2e46d] LinearAlgebra


In [2]:
#using Pkg; Pkg.add("LinearAlgebra"); Pkg.add("Markdown")
# using Documenter
using LinearAlgebra
using Markdown                             # Pour que les docstrings en début des fonctions ne posent
                                           # pas de soucis. Ces docstrings sont utiles pour générer 
                                           # la documentation sous GitHub

tol_erreur = sqrt(eps())

include("Algorithme_De_Newton.jl")

include("../test/fonctions_de_tests.jl")

include("../test/tester_algo_newton.jl")

# Affichage les sorties de l'algorithme des Régions de confiance
function my_afficher_resultats(algo,nom_fct,point_init,xmin,fxmin,flag,sol_exacte,nbiters)
	println("-------------------------------------------------------------------------")
	printstyled("Résultats de : ",algo, " appliqué à ",nom_fct, " au point initial ", point_init, ":\n",bold=true,color=:blue)
	println("  * xsol = ",xmin)
	println("  * f(xsol) = ",fxmin)
	println("  * nb_iters = ",nbiters)
	println("  * flag = ",flag)
	println("  * sol_exacte : ", sol_exacte)
end

# Fonction f0
# -----------
f0(x) =  sin(x)
# la gradient de la fonction f0
grad_f0(x) = cos(x)
# la hessienne de la fonction f0
hess_f0(x) = -sin(x)
sol_exacte = -pi/2
options = []

x0 = sol_exacte
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f0,grad_f0,hess_f0,x0,options)
my_afficher_resultats("Newton","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = -pi/2+0.5
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f0,grad_f0,hess_f0,x0,options)
my_afficher_resultats("Newton","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = pi/2
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f0,grad_f0,hess_f0,x0,options)
my_afficher_resultats("Newton","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)

# autres tests de l'annexe 
# Tester l'algorithme de Newton
	tester_algo_newton(true,Algorithme_De_Newton)

-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0 au point initial -1.5707963267948966:
  * xsol = -1.5707963267948966
  * f(xsol) = -1.0
  * nb_iters = 0
  * flag = 0
  * sol_exacte : -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0 au point initial -1.0707963267948966:
  * xsol = -1.5707963267949088
  * f(xsol) = -1.0
  * nb_iters = 3
  * flag = 0
  * sol_exacte : -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0 au point initial 1.5707963267948966:
  * xsol = 1.5707963267948966
  * f(xsol) = 1.0
  * nb_iters = 0
  * flag = 0
  * sol_exacte : -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : algorithme de Newton  appliqué à fct1 au point initial solution :
  * xsol = [1, 1, 1]
  * f(xsol) = 0
  * nb_iters = 0

LoadError: Some tests did not pass: 13 passed, 1 failed, 0 errored, 0 broken.

## Interprétation 

Justifier

1. les résultats obtenus pour l'exemple $f_0$ ci-dessus;

2. que l’algorithme implémenté converge en une itération pour $f_{1}$;

3. que l’algorithme puisse ne pas converger pour $f_{2}$ avec certains points initiaux.

## Réponses
1.    * Dans le premier cas, l'algorithme utilise pour valeur initiale de $x_{0}$ celle de la solution exacte, le gradient est donc nul, la solution est donnée directement par l'algorithme. 
    * Dans le deuxième cas, l'algorithme fait trois itérations avant de trouver une solution proche de la solution exacte, car la valeur initiale donnée est proche de la solution. 
    * Dans le troisème cas, l'algorithme renvoie une solution directement car la valeur donnée correspond à un maximum de la fonction, qui annule aussi le gradient, mais qui n'est donc pas la solution cherchée pour le minimum.
<br/>
<br/>
2. Pour $f_{1}$, les points initiaux donnés permettent une convergence de la solution vers des points proches de la solution réels. Les calculs du gradient font osciller vers la vraie solution. 
<br/>
<br/>
3.   * Pour $f_{2}$, les tests pour les valeurs de départ qui sont la solution, $x_{21}$ et $x_{22}$ permettent d'obtenir une solution, car ces valeurs ne sont pas trop éloignées de la solution réelle du problème. 
    * Pour la valeur de départ $x_{23}$, la solution étant plus éloignée de la valeur initiale, et celle-ci étant proche de zéro, l'évolution de $x_k$ est plus lente et oscille davantage autour de la solution, donc l'algorithme ne converge pas et l'on en sort car le nombre d'itération est trop élevé.



# Régions de confiance avec pas de cauchy 

## Implémentation 

1. Coder l'algorithme du pas de Cauchy d’un sous-problème de
régions de confiance (fichier `Pas_De_Cauchy.jl`). Tester sur les quadratiques proposées en Annexe B.

2. Coder l'algorithme de régions de confiance (fichier `Regions_De_Confiance.jl`). Tester sur les problèmes de l’Annexe A.

In [3]:
include("Gradient_Conjugue_Tronque.jl")
include("Regions_De_Confiance.jl")
tol_erreur = sqrt(eps())
# test de l'annexe

include("../test/tester_pas_de_cauchy.jl")
include("../test/tester_regions_de_confiance_cauchy.jl")

# Tester l'algorithme du pas de Cauchy
	tester_pas_de_cauchy(true,Pas_De_Cauchy)

# Tester l'algorithme des Régions de confiance avec PasdeCauchy
	tester_regions_de_confiance(true,Regions_De_Confiance)


Cauchy 2 = [-0.9230769230769232, -0.30769230769230776]
Cauchy 3 = [-0.8538149682454623, -0.2846049894151541]
Cauchy 4 = [5.000000000000001, -2.5000000000000004]
Cauchy 5= [4.47213595499958, -2.23606797749979]
Cauchy 6= [-4.743416490252569, -1.5811388300841895]
Cauchy 7= [-2.23606797749979, -4.47213595499958]
Test Summary: | Pass  Total
Pas de Cauchy |    7      7
-------------------------------------------------------------------------
Résultats de : régions de confiance avec cauchy appliqué à fonction 1 au point initial x011 :
  * xsol = [1.0000000310867154, 0.9999999957837182, 0.999999960480721]
  * f(xsol) = 2.812589785349316e-15
  * nb_iters = 48
  * flag = 0
  * sol_exacte : [1, 1, 1]
-------------------------------------------------------------------------
Résultats de : régions de confiance avec cauchy appliqué à fonction 1 au point initial x012 :
  * xsol = [1.0000001114825787, 0.9999999912691591, 0.9999998710557396]
  * f(xsol) = 3.027462898391515e-14
  * nb_iters = 42
  * fla

Test.DefaultTestSet("La méthode des RC ", Any[Test.DefaultTestSet("avec Cauchy ", Any[], 5, false, false)], 0, false, false)

## Interprétation 

1. Quelle relation lie la fonction test $f_1$ et son modèle de Taylor à l’ordre 2 ? Comparer alors les performances de Newton et RC-Pas de Cauchy sur cette fonction.

2. Le rayon initial de la région de confiance est un paramètre important dans l’analyse
de la performance de l’algorithme. Sur quel(s) autre(s) paramètre(s) peut-on jouer
pour essayer d’améliorer cette performance ? Étudier l’influence d’au moins deux de
ces paramètres.

## Réponses

1. La série de Taylor à l'ordre 2 de la fonction $f_1$ nous donne pour approximation de $f_1$ pour $s$ petit, pour $x \in R^n$, $f(x+s) = f(x) + (\nabla(f)(x))^Ts + (1/2)s^T\nabla^2(f)(x)s$. Le pas de Cauchy permet donc de considérer en plus du gradient les autres directions autour de $x$ tout en restant proche de la valeur de $x$, car l'on reste à une distance $\delta_k$ de $x$ dans cette méthode. Les performances de l'algorithme de Newton sont moins efficaces, car si le gradient coupe l'axe des abcisses en un point trop éloigné de la solution du problème de minimisation, l'algorithme de Newton diverge (comme pour $f_2$ pour certains points initiaux). Le pas de Cauchy permet de rester aux alentours de la solution du problème en donnant une évolution de $x_k$ plus fine.
<br/>
<br/>
2. L'évolution de $\delta_k$ dépend des valeurs de $\gamma_1$ et $\gamma_2$, qui permettent de réduire ou d'augmenter la région de confiance selon l'approximation de $f$ par sa série de Taylor, et la perfomance dépend aussi de $\eta_1$ et $\eta_2$ qui permettent de déterminer si l'approximation de $f$ par sa série de Taylor est assez bonne ou non. 
    - En modifiant le paramètre $\eta_1$ on observe qu'avec un $\eta_1$ plus élevé, la solution trouvée se rapproche davantage de la solution exacte du problème pour un même nombre d'itération. Pour $f_2$ au point initial $x_{023}$, on trouve une solution de $[0.9999370949654659, 0.9998739880304974]$ avec $\eta_1 = 0.1$, et une solution de $[0.9999993488201874, 0.9999986988623017]$ avec $\eta_1 = 0.5$, ce qui rapproche $f_2$ de $0$ à la solution trouvée de $10^{-4}$ environ. L'algorithme converge donc plus rapidement. 
    - En modifiant le pramètre $\gamma_2$ on peut observer qu'en augmentant sa valeur, la solution trouvée est moins proche de la solution exact pour un même nombre d'itération. En effet pour $f_2$ au point initial $x_{023}$, pour $\gamma_2 = 2$ on obtient une solution de $[0.9999370949654659, 0.9998739880304974]$, et avec $\gamma_2 = 5$, on obtient pour solution $[0.9987471516481539, 0.9974937305838003]$, ce qui éloigne $f_2$ de $0$ à la solution trouvée de $10^{-3}$. L'algorithme converge donc moins rapidement. 


# Régions de confiance avec gradient conjugué tronqué

## Implémentation 

1. Implémenter l’algorithme du Gradient Conjugué Tronqué, en se basant sur le cours (fichier `Gradient_Conjugue_Tronque.jl`).
On validera les résultats sur les fonctions de l’Annexe C.

2. Intégrer finalement l’algorithme du Gradient Conjugué Tronqué dans le code de
régions de confiance, et appliquer ce code pour résoudre les exemples proposés en
Annexe A.

In [4]:
include("Gradient_Conjugue_Tronque.jl")
include("Regions_De_Confiance.jl")
tol_erreur = sqrt(eps())

include("../test/tester_gct.jl")
include("../test/tester_regions_de_confiance_gct.jl")


# Tester l'algorithme du gradient conjugué tronqué
	tester_gct(true,Gradient_Conjugue_Tronque)

# Tester l'algorithme des Régions de confiance avec GCT
	tester_regions_de_confiance(true,Regions_De_Confiance)


Test Summary: | Pass  Total
Gradient-CT   |   11     11
-------------------------------------------------------------------------
Résultats de : régions de confiance avec gct appliqué à fonction 1 au point initial x011 :
  * xsol = [1.0000000000000002, 1.0000000000000002, 1.0000000000000004]
  * f(xsol) = 1.627025617018337e-30
  * nb_iters = 1
  * flag = 0
  * sol_exacte : [1, 1, 1]
-------------------------------------------------------------------------
Résultats de : régions de confiance avec gct appliqué à fonction 1 au point initial x012 :
  * xsol = [1.0, 1.000000000000001, 1.0000000000000022]
  * f(xsol) = 2.179228250673045e-29
  * nb_iters = 2
  * flag = 0
  * sol_exacte : [1, 1, 1]
-------------------------------------------------------------------------
Résultats de : régions de confiance avec gct appliqué à fonction 2 au point initial x021 :
  * xsol = [0.9999999936622372, 0.9999999872742869]
  * f(xsol) = 4.041911662879635e-17
  * nb_iters = 38
  * flag = 0
  * sol_exacte :

Test.DefaultTestSet("La méthode des RC ", Any[Test.DefaultTestSet("avec GCT ", Any[], 5, false, false)], 0, false, false)

## Interprétation  

1. Comparer la décroissance obtenue avec celle du pas de Cauchy, en retournant, dans
un premier temps le dernier itéré admissible à courbure positive (c’est à dire, que si
l’une ou l’autre des deux conditions (b) ou (d) sont rencontrées dans l’algorithme 3,
alors on ne calcule pas ``σ_{j}`` et on retourne le dernier itéré ``s_{j}`` directement).

2. Comparer la décroissance obtenue avec celle du pas de Cauchy, en imposant la sortie
dans l’algorithme 3 au bout d’une itération seulement. Que remarquez vous ?

3. Comparer la décroissance obtenue avec celle du pas de Cauchy dans le cas général.

4. Quels sont les avantages et inconvénients des deux approches ?

## Réponses
1. On trouve que les derniers itérés pour l'algorithme 3 sont les mêmes que les solutions renvoyées par le pas de Cauchy.
<br/>
<br/>
2. En imposant une sortie en une itération, on trouve avec l'algorithme 3 les mêmes solutions que pour le pas de Cauchy lorsque l'on rencontre les conditions b ou d, et des solutions proches au dixième près de celle obtenues par le pas de Cauchy pour les autres conditions. 
<br/>
<br/>
3. Les deux algorithmes permettent en une itération d'obtenir des résultats similaires, mais en calculant la valeur de la fonction à minimiser q en la solution trouvée pour chaque algorithme, on observe que le minimum calculé avec le gradient conjugué tronqué sans limiter à une itération la boucle est plus bas que celui obtenu par le pas de Cauchy. 
<br/>
<br/>
4. La première approche avec le pas de Cauchy est plus efficace en temps de calcul que celle du gradient conjugué tronqué car c'est un calcul direct, mais la seconde approche est plus précise dans le calcul de la solution s renvoyée. Le pas entre $x_k$ et $x_{k+1}$ sera donc plus rapide à calculer en temps avec la première approche, mais la précision de l'évolution avec la deuxième approche permettra de trouver la solution par l'algorithme de région de confiance plus rapidement, le nombre d'itération dans la boucle de l'algorithme de région de confiance est plus faible avec la deuxième méthode. Les deux méthodes ont donc chacune leurs avantages par rapport au temps de calcul sur l'algortihme global. 


# Lagrangien augmenté

## Implémentation
1.Choisir des critères d’arrêt pour la convergence de l'algorithme.

2.Implémenter l'algorithme du lagrangien augmenté, en utilisant les différentes méthodes
qui ont été vues en première partie pour la résolution de la suite de problémes sans
contraintes (fichier `Lagrangien_Augmente.jl`)
 
3.Tester les différentes variantes sur les problèmes en Annexe D.
 

In [5]:

include("Algorithme_De_Newton.jl")
include("Gradient_Conjugue_Tronque.jl")
include("Lagrangien_Augmente.jl")
include("Pas_De_Cauchy.jl")
include("Regions_De_Confiance.jl")
tol_erreur = sqrt(eps())

include("../test/tester_lagrangien_augmente.jl")


# Tester l'algorithme du Lagrangien Augmenté
  	tester_lagrangien_augmente(true,Lagrangien_Augmente)

-------------------------------------------------------------------------
Résultats de : Lagrangien augmenté avec newton appliqué à fonction 1 au point initial x01 :
  * xsol = [0.5000000000165669, 1.2499999999917166, 0.5000000000165669]
  * f(xsol) = 2.2499999998508984
  * nb_iters = 9
  * flag = 0
  * sol_exacte : [0.5, 1.25, 0.5]
-------------------------------------------------------------------------
Résultats de : Lagrangien augmenté avec newton appliqué à fonction 1 au point initial x02 :
  * xsol = [0.5000000126442301, 1.249999993677885, 0.5000000126442301]
  * f(xsol) = 2.2499998862019313
  * nb_iters = 6
  * flag = 0
  * sol_exacte : [0.5, 1.25, 0.5]
-------------------------------------------------------------------------
Résultats de : Lagrangien augmenté avec newton appliqué à fonction 2 au point initial x03 :
  * xsol = [0.9072339234314191, 0.8227553889173606]
  * f(xsol) = 0.008615657546839645
  * nb_iters = 5
  * flag = 0
  * sol_exacte : [0.9072339605110892, 0.82275545

Test.DefaultTestSet("Lagrangien augmenté  ", Any[Test.DefaultTestSet("Avec newton", Any[], 4, false, false), Test.DefaultTestSet("Avec gct", Any[], 4, false, false), Test.DefaultTestSet("Avec cauchy", Any[], 4, false, false)], 0, false, false)

## Interprétation
 1.Commenter les résultats obtenus, en étudiant notamment les valeurs de $\lambda_k$ et $\mu_k$.
 
 2.Étudier l'influence du paramètre $\tau$ dans la performance de l'algorithme.
 
 3.**Supplémentaire** : 
      Que proposez-vous comme méthode pour la résolution des problèmes avec
      des contraintes à la fois d'égalité et d'inégalité ? Implémenter (si le temps le permet)
      ce nouvel algorithme


## Réponses
1. Les solutions trouvées sont proches des solutions exactes et sont données en un nombre d'itération relativement faible. La valeur de $\lambda_k$ change de façon remarquable selon la fonction que l'on étudie, elle est d'environ $4.5$ pour la fonction $f_1$, et d'environ $3.86e-2$ pour la fonction $f_2$, tandis que celle de $\mu_k$ dépend du point initial que l'on prend pour lancer l'algorithme, et cette valeur est exactement la même lorsque l'on change d'algorithme pour le calcul du minimum du problème sans contraintes dans la boucle de l'algorithme principal. 
<br/><br/>
2. Lorsque l'on augmente la valeur de $\tau$, le nombre d'itération nécessaire pour trouver la solution diminue, et la valeur de $\mu_k$ augmente. En effet la valeur de $\tau$ est utilisée dans le point c de l'algorithme pour mettre à jour la valeur de $\mu_k$ puis des autres paramètres lorsque la solution trouvée pour minimiser le sous problème sans contraintes ne vérifie pas les contraintes du problème global.
<br/><br/>
3. Un problème avec des contraintes d'égalilté et d'inégalité peut s'écrire :
<br/>
    $(P) :$ $min_{x \in R^n}$ $f(x)$ <br/> $c_e(x) = 0$ <br/> $c_i(x) \le 0 $
    <br/> <br/>
    On peut donc écrire les contraintes d'inégalité comme : <br/>
    $c_i(x) + s = 0$ , $s \ge 0$ <br/>
    On a alors une nouvelle contrainte qui devient une égalité dans notre problème, et l'on doit considérer en plus dans la fonction $L_A$ de l'algorithme la condition $s \ge 0$ qui se traduit par une barrière logarithmique proportionnelle à $\sum_{i=1}^{n} log(s_i)$ qui doit être soustraite de la fonction $L_A$, ainsi que l'ajout aux contraintes déjà présentes de la contrainte $c : x \to c_i + s$. <br/>
    
    
